File: src/task-bazaar.js
/**
 * TaskBazaar - 任务市场
 * 发布任务、竞标、奖励分配
 */

const EventEmitter = require('events');
const crypto = require('crypto');

class TaskBazaar extends EventEmitter {
    constructor(options = {}) {
        super();
        this.nodeId = options.nodeId;
        this.memoryStore = options.memoryStore;
        this.ledger = options.ledger || null;
        this.walletAccountId = options.walletAccountId || null;
        this.ratingStore = options.ratingStore || null;
        this.dataDir = options.dataDir || process.cwd();
        this.tasksPath = require('path').join(this.dataDir, 'tasks.json');
        
        this.tasks = new Map(); // taskId -> task
        this.submissions = new Map(); // taskId -> [solutions]
        this.completedTasks = new Set();
        this.loadFromDisk();
    }

    loadFromDisk() {
        const fs = require('fs');
        if (!fs.existsSync(this.tasksPath)) return;
        try {
            const raw = JSON.parse(fs.readFileSync(this.tasksPath, 'utf8'));
            if (Array.isArray(raw)) {
                raw.forEach(t => {
                    if (t && t.taskId) {
                        this.tasks.set(t.taskId, t);
                        if (t.status === 'completed') {
                            this.completedTasks.add(t.taskId);
                        }
                    }
                });
            }
        } catch (e) {
        }
    }

    saveToDisk() {
        const fs = require('fs');
        try {
            const payload = Array.from(this.tasks.values());
            fs.writeFileSync(this.tasksPath, JSON.stringify(payload, null, 2));
        } catch (e) {
        }
    }
    
    // 发布任务
    async publishTask(task) {
        // 验证任务
        if (!task.description || !task.bounty) {
            throw new Error('Invalid task: missing description or bounty');
        }
        
        // 生成 taskId
        if (!task.taskId) {
            task.taskId = this.generateTaskId(task);
        }

        task.publisher = task.publisher || this.nodeId;
        task.bounty.token = task.bounty.token || 'CLAW';

        task.escrowAccountId = task.escrowAccountId || this.getEscrowAccountId(task.taskId);
        
        // 存储任务
        task.status = this.isEscrowFunded(task) ? 'open' : 'pending_escrow';
        task.submissions = [];
        this.tasks.set(task.taskId, task);
        this.saveToDisk();
        
        console.log(`🎯 Task published: ${task.taskId}`);
        console.log(`   Bounty: ${task.bounty.amount} ${task.bounty.token}`);
        
        this.emit('task:published', task);
        return task.taskId;
    }
    
    // 处理网络中的新任务
    async handleNewTask(task) {
        if (this.tasks.has(task.taskId)) {
            return; // 已存在
        }
        task.escrowAccountId = task.escrowAccountId || this.getEscrowAccountId(task.taskId);
        task.status = this.isEscrowFunded(task) ? 'open' : 'pending_escrow';
        task.submissions = [];
        this.tasks.set(task.taskId, task);
        this.saveToDisk();
        
        console.log(`📬 New task from network: ${task.taskId}`);
        this.emit('task:received', task);
    }
    
    // 提交解决方案
    async submitSolution(taskId, solution, solverId) {
        const task = this.tasks.get(taskId);
        if (!task) {
            throw new Error('Task not found');
        }
        
        if (task.status !== 'open') {
            throw new Error('Task is not open');
        }
        
        if (this.completedTasks.has(taskId)) {
            return { success: false, reason: 'Task already completed' };
        }
        
        // 验证解决方案（简化版）
        const valid = await this.validateSolution(task, solution);
        
        if (valid) {
            // 第一个有效解获胜
            this.completedTasks.add(taskId);
            task.status = 'completed';
            task.winner = solverId;
            
            const reward = task.bounty?.amount || 0;
            
            console.log(`🏆 Task completed: ${taskId}`);
            console.log(`   Winner: ${solverId}`);
            console.log(`   Reward: ${reward}`);
            
            this.emit('task:completed', {
                taskId,
                winner: solverId,
                reward
            });
            
            return {
                success: true,
                winner: true,
                reward,
                winnerId: solverId
            };
        }
        
        return {
            success: false,
            reason: 'Invalid solution'
        };
    }
    
    // 验证解决方案
    async validateSolution(task, solution) {
        // 简化验证：检查是否有代码和说明
        if (!solution.code && !solution.description) {
            return false;
        }
        
        // 如果是代码任务，尝试执行验证
        if (task.type === 'code' && solution.code) {
            try {
                // 这里简化处理，实际应该沙箱执行
                return solution.code.length > 10; // 至少10字符
            } catch (e) {
                return false;
            }
        }
        
        return true;
    }
    
    // 获取任务列表
    getTasks(filter = {}) {
        let tasks = Array.from(this.tasks.values());
        
        if (filter.status) {
            tasks = tasks.filter(t => t.status === filter.status);
        }
        
        if (filter.type) {
            tasks = tasks.filter(t => t.type === filter.type);
        }
        
        return tasks.sort((a, b) => new Date(b.published_at) - new Date(a.published_at));
    }
    
    // 获取单个任务
    getTask(taskId) {
        return this.tasks.get(taskId);
    }
    
    // 获取任务数量
    getTaskCount() {
        return this.tasks.size;
    }
    
    // 更新任务状态
    updateTask(taskId, updates) {
        const task = this.tasks.get(taskId);
        if (task) {
            this.tasks.set(taskId, { ...task, ...updates });
            this.saveToDisk();
            return this.tasks.get(taskId);
        }
        return null;
    }
    
    // 完成任务
    completeTask(taskId, result) {
        const task = this.tasks.get(taskId);
        if (task) {
            task.status = 'completed';
            task.completedAt = new Date().toISOString();
            task.result = result;
            this.tasks.set(taskId, task);
            this.saveToDisk();
            return task;
        }
        return null;
    }
    
    // 存入积分
    deposit(amount) {
        this.balance += amount;
        return this.balance;
    }
    
    // 获取余额
    getBalance() {
        if (!this.ledger || !this.walletAccountId) {
            return { available: 0, locked: 0 };
        }
        const available = this.ledger.getBalance(this.walletAccountId);
        let locked = 0;
        for (const task of this.tasks.values()) {
            if (task.publisher === this.nodeId && task.escrowAccountId && task.status !== 'completed') {
                locked += this.ledger.getBalance(task.escrowAccountId) || 0;
            }
        }
        return { available, locked };
    }

    isNodeAllowed(nodeId) {
        if (!this.ratingStore || !nodeId) return true;
        return !this.ratingStore.isDisqualified(nodeId);
    }

    isEscrowFunded(task) {
        if (!this.ledger || !task?.escrowAccountId) return true;
        const balance = this.ledger.getBalance(task.escrowAccountId);
        return Number(balance) >= Number(task.bounty?.amount || 0);
    }

    tryActivatePendingTasks() {
        for (const task of this.tasks.values()) {
            if (task.status === 'pending_escrow' && this.isEscrowFunded(task)) {
                task.status = 'open';
                this.tasks.set(task.taskId, task);
            }
        }
    }

    getEscrowAccountId(taskId) {
        const hash = crypto.createHash('sha256').update(String(taskId)).digest('hex').slice(0, 24);
        return `escrow_${hash}`;
    }
    
    // 创建Swarm任务（复杂任务分解）
    async createSwarmTask(description, subtasks, totalBounty) {
        const swarmTask = {
            taskId: this.generateTaskId(),
            type: 'swarm',
            description,
            status: 'open',
            subtasks: subtasks.map((st, i) => ({
                id: `sub_${i}`,
                description: st.description,
                weight: st.weight || 1,
                status: 'open',
                reward: (totalBounty * 0.85) * (st.weight / subtasks.reduce((a, s) => a + s.weight, 0))
            })),
            proposerReward: totalBounty * 0.05,
            aggregatorReward: totalBounty * 0.10,
            bounty: {
                amount: totalBounty,
                token: 'CLAW'
            },
            published_at: new Date().toISOString()
        };
        
        return await this.publishTask(swarmTask);
    }
    
    generateTaskId() {
        return 'task_' + crypto.randomBytes(8).toString('hex');
    }
    
    // 获取统计
    getStats() {
        const tasks = Array.from(this.tasks.values());
        return {
            total: tasks.length,
            open: tasks.filter(t => t.status === 'open').length,
            completed: tasks.filter(t => t.status === 'completed').length,
            totalRewards: tasks
                .filter(t => t.status === 'completed')
                .reduce((sum, t) => sum + (t.bounty?.amount || 0), 0)
        };
    }
}

module.exports = TaskBazaar;

File: src/task_bazaar.rs
use crate::store::Store;
use crate::util::now_iso;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use tokio::sync::Mutex;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TaskBounty {
    pub amount: i64,
    pub token: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TaskBid {
    pub node_id: String,
    pub amount: i64,
    pub timestamp: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Task {
    pub task_id: String,
    pub description: String,
    #[serde(rename = "type")]
    pub task_type: Option<String>,
    pub bounty: TaskBounty,
    pub tags: Vec<String>,
    pub publisher: String,
    pub status: String,
    pub submissions: Vec<serde_json::Value>,
    pub bids: Vec<TaskBid>,
    #[serde(alias = "published_at")]
    pub published_at: String,
    #[serde(alias = "voting_started_at")]
    pub voting_started_at: Option<i64>,
    #[serde(alias = "assigned_to")]
    pub assigned_to: Option<String>,
    #[serde(alias = "assigned_at")]
    pub assigned_at: Option<i64>,
    pub winner: Option<String>,
    #[serde(alias = "completed_at")]
    pub completed_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskStats {
    pub total: usize,
    pub open: usize,
    pub completed: usize,
    pub total_rewards: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BalanceStats {
    pub available: i64,
    pub locked: i64,
}

pub struct TaskBazaar {
    pub node_id: String,
    store: Arc<Mutex<Store>>,
    tasks: HashMap<String, Task>,
    completed_tasks: HashSet<String>,
}

impl TaskBazaar {
    pub fn new(node_id: String, store: Arc<Mutex<Store>>) -> Self {
        Self {
            node_id,
            store,
            tasks: HashMap::new(),
            completed_tasks: HashSet::new(),
        }
    }

    pub async fn publish_task(&mut self, mut task: Task) -> Result<String, String> {
        if task.description.trim().is_empty() {
            return Err("Invalid task: missing description".to_string());
        }
        if task.bounty.amount <= 0 {
            return Err("Invalid task: missing bounty".to_string());
        }
        if task.task_id.is_empty() {
            task.task_id = format!("task_{}", crate::util::random_hex(8));
        }
        if task.publisher.is_empty() {
            task.publisher = self.node_id.clone();
        }
        if task.bounty.token.is_empty() {
            task.bounty.token = "CLAW".to_string();
        }
        let published_at = now_iso();
        task.published_at = published_at;
        task.status = "open".to_string();
        task.submissions = Vec::new();
        task.bids = Vec::new();

        let mut store = self.store.lock().await;
        let publisher_account = store.ensure_account(&task.publisher, "gep-lite-v1")?;
        store.lock_escrow(&task.task_id, &publisher_account.account_id, task.bounty.amount, &task.bounty.token)?;

        let task_id = task.task_id.clone();
        self.tasks.insert(task_id.clone(), task);
        Ok(task_id)
    }

    pub async fn handle_new_task(&mut self, mut task: Task) {
        if self.tasks.contains_key(&task.task_id) {
            return;
        }
        task.status = "open".to_string();
        task.submissions = Vec::new();
        task.bids = Vec::new();
        self.tasks.insert(task.task_id.clone(), task);
    }

    pub async fn submit_solution(
        &mut self,
        task_id: &str,
        solution: serde_json::Value,
        solver_node_id: &str,
    ) -> Result<serde_json::Value, String> {
        let task = self.tasks.get_mut(task_id).ok_or("Task not found")?;
        if task.status != "open" && task.status != "assigned" {
            return Err("Task is not open".to_string());
        }
        if self.completed_tasks.contains(task_id) {
            return Ok(serde_json::json!({ "success": false, "reason": "Task already completed" }));
        }
        let valid = Self::validate_solution(task, &solution);
        if !valid {
            return Ok(serde_json::json!({ "success": false, "reason": "Invalid solution" }));
        }
        self.completed_tasks.insert(task_id.to_string());
        task.status = "completed".to_string();
        task.winner = Some(solver_node_id.to_string());
        task.completed_at = Some(now_iso());

        let mut store = self.store.lock().await;
        let winner_account = store.ensure_account(solver_node_id, "gep-lite-v1")?;
        let reward = store.release_escrow(task_id, &winner_account.account_id)?;
        Ok(serde_json::json!({ "success": true, "winner": true, "reward": reward }))
    }

    pub fn update_task(&mut self, task_id: &str, updates: serde_json::Value) -> Option<Task> {
        let task = self.tasks.get_mut(task_id)?;
        if let Some(status) = updates.get("status").and_then(|v| v.as_str()) {
            task.status = status.to_string();
        }
        if let Some(bids) = updates.get("bids").and_then(|v| v.as_array()) {
            task.bids = bids
                .iter()
                .filter_map(|b| serde_json::from_value::<TaskBid>(b.clone()).ok())
                .collect();
        }
        if let Some(assigned) = updates.get("assigned_to").and_then(|v| v.as_str()) {
            task.assigned_to = Some(assigned.to_string());
        }
        if let Some(at) = updates.get("assigned_at").and_then(|v| v.as_i64()) {
            task.assigned_at = Some(at);
        }
        if let Some(vote) = updates.get("voting_started_at").and_then(|v| v.as_i64()) {
            task.voting_started_at = Some(vote);
        }
        Some(task.clone())
    }

    pub fn get_tasks(&self) -> Vec<Task> {
        let mut tasks: Vec<Task> = self.tasks.values().cloned().collect();
        tasks.sort_by(|a, b| b.published_at.cmp(&a.published_at));
        tasks
    }

    pub fn get_task(&self, task_id: &str) -> Option<Task> {
        self.tasks.get(task_id).cloned()
    }

    pub fn get_task_count(&self) -> usize {
        self.tasks.len()
    }

    pub fn get_stats(&self) -> TaskStats {
        let tasks: Vec<Task> = self.tasks.values().cloned().collect();
        TaskStats {
            total: tasks.len(),
            open: tasks.iter().filter(|t| t.status == "open").count(),
            completed: tasks.iter().filter(|t| t.status == "completed").count(),
            total_rewards: tasks
                .iter()
                .filter(|t| t.status == "completed")
                .map(|t| t.bounty.amount)
                .sum(),
        }
    }

    pub async fn get_balance(&self) -> Result<BalanceStats, String> {
        let store = self.store.lock().await;
        let available = store.get_balance(&self.node_id)?;
        let node_account_id = store.get_account_id_by_node(&self.node_id)?.unwrap_or_default();
        let locked = store
            .list_escrows()?
            .into_iter()
            .filter(|e| e.from_account_id == node_account_id)
            .map(|e| e.amount)
            .sum();
        Ok(BalanceStats { available, locked })
    }

    pub fn add_bid(&mut self, task_id: &str, bid: TaskBid) -> Option<Task> {
        let task = self.tasks.get_mut(task_id)?;
        if task.bids.iter().any(|b| b.node_id == bid.node_id) {
            return Some(task.clone());
        }
        task.bids.push(bid);
        if task.status == "open" {
            task.status = "voting".to_string();
            task.voting_started_at = Some(chrono::Utc::now().timestamp_millis());
        }
        Some(task.clone())
    }

    pub fn determine_winner(&self, task: &Task) -> Option<TaskBid> {
        if task.bids.is_empty() {
            return None;
        }
        let mut bids = task.bids.clone();
        bids.sort_by(|a, b| {
            if a.amount != b.amount {
                a.amount.cmp(&b.amount)
            } else {
                a.timestamp.cmp(&b.timestamp)
            }
        });
        bids.first().cloned()
    }

    fn validate_solution(task: &Task, solution: &serde_json::Value) -> bool {
        if solution.get("code").is_none() && solution.get("description").is_none() {
            return false;
        }
        if let Some(task_type) = &task.task_type {
            if task_type == "code" {
                if let Some(code) = solution.get("code").and_then(|v| v.as_str()) {
                    return code.len() > 10;
                }
            }
        }
        true
    }
}
